File: core-python/src/sentinel_dispatch/domain/routing/heuristica.py
"""Heurística Haversine para A* sobre el grafo vial.

Distancia ortodrómica entre dos coordenadas geográficas, escalada a tiempo
de viaje por la velocidad máxima del sistema. Es **admisible** (nunca
sobreestima el costo real) porque ningún tramo del grafo se recorre a
velocidad efectiva superior a :data:`V_MAX_MS` (SRS sec. 2.6-B).

Lógica pura: depende solo de math.stdlib.
"""

from __future__ import annotations

import math

V_MAX_KMH: float = 180.0
"""Velocidad máxima absoluta del sistema en km/h.

Cota superior sobre la velocidad efectiva de cualquier arista bajo
cualquier combinación realista de ``factor_hora * factor_sirena``.
Garantiza la admisibilidad de :func:`haversine_segundos` (la heurística
nunca sobreestima el costo real).

Cálculo del techo: arista más rápida en la tabla Chile (ADR-0010 §2) =
``motorway = 120 km/h``. Factor combinado más rápido = ``factor_hora *
factor_sirena`` con ``factor_hora <= 1`` (1.0 es hora valle; valores
menores ralentizan por tráfico) y ``factor_sirena = 1.4`` (SRS sec.
2.6-B, sirena activa). Pico real ``120 * 1.0 * 1.4 = 168 km/h``. Se
adopta ``V_MAX = 180`` con margen del 7 % para absorber outliers del
grafo OSM o futuras variaciones de la tabla horaria.
"""

V_MAX_MS: float = V_MAX_KMH * 1000.0 / 3600.0
"""Velocidad máxima del sistema en m/s (≈ 38.89 m/s)."""

RADIO_TIERRA_M: float = 6_371_000.0
"""Radio medio de la Tierra en metros (esfera ideal, error ≤ 0.5%)."""


def haversine_m(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Distancia ortodrómica entre dos coordenadas, en metros.

    Fórmula clásica de Haversine sobre esfera de radio :data:`RADIO_TIERRA_M`.
    El error frente al modelo elipsoidal (WGS84) es < 0.5%, despreciable
    para el orden de magnitud del proyecto.
    """
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    dlat = lat2_rad - lat1_rad
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2.0) ** 2
        + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2.0) ** 2
    )
    c = 2.0 * math.asin(math.sqrt(a))
    return RADIO_TIERRA_M * c


def haversine_segundos(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Heurística admisible h(n) para el A*: tiempo mínimo posible en segundos.

    Calcula la distancia ortodrómica y la divide por :data:`V_MAX_MS`.
    Como ninguna arista del grafo puede recorrerse a más de ``V_MAX``,
    ``h(n) ≤ costo_real(n, objetivo)`` para todo ``n`` — la condición
    de admisibilidad que A* exige para garantizar optimalidad.
    """
    return haversine_m(lat1, lon1, lat2, lon2) / V_MAX_MS

File: core-java/src/main/java/cl/ucen/sentinel/domain/routing/Heuristica.java
package cl.ucen.sentinel.domain.routing;

/**
 * Heurística Haversine admisible para el A* sobre el grafo vial (ADR-0010 §1).
 *
 * <p>Espejo de {@code sentinel_dispatch/domain/routing/heuristica.py}. Calcula la distancia
 * ortodrómica entre dos nodos del grafo y la divide por la velocidad máxima del sistema,
 * produciendo un tiempo mínimo posible en segundos. La heurística es admisible: ninguna arista del
 * grafo puede recorrerse a más de {@link #V_MAX_KMH}, por lo que {@code h(n) ≤ costo_real(n,
 * destino)} para todo {@code n}.
 *
 * <p>Lógica pura: depende solo de {@code java.lang.Math} y del puerto {@link GrafoVial}.
 */
public final class Heuristica {

  /** Velocidad máxima absoluta del sistema en km/h. Espejo de {@code V_MAX_KMH = 180.0} Python. */
  public static final double V_MAX_KMH = 180.0;

  /** Velocidad máxima en m/s (≈ 38.89 m/s). Espejo de {@code V_MAX_MS} Python. */
  public static final double V_MAX_MS = V_MAX_KMH * 1000.0 / 3600.0;

  /** Radio medio de la Tierra en metros. Espejo de {@code RADIO_TIERRA_M = 6_371_000.0} Python. */
  public static final double RADIO_TIERRA_M = 6_371_000.0;

  private Heuristica() {}

  /**
   * Distancia ortodrómica entre dos puntos, en metros.
   *
   * <p>Espejo de {@code haversine_m(lat1, lon1, lat2, lon2)} Python. Fórmula Haversine clásica
   * sobre esfera de radio {@link #RADIO_TIERRA_M}.
   *
   * @param lat1 latitud del punto 1 en grados decimales
   * @param lon1 longitud del punto 1 en grados decimales
   * @param lat2 latitud del punto 2 en grados decimales
   * @param lon2 longitud del punto 2 en grados decimales
   * @return distancia en metros
   */
  public static double haversineM(double lat1, double lon1, double lat2, double lon2) {
    double lat1Rad = Math.toRadians(lat1);
    double lat2Rad = Math.toRadians(lat2);
    double dlat = lat2Rad - lat1Rad;
    double dlon = Math.toRadians(lon2 - lon1);
    double a =
        Math.sin(dlat / 2.0) * Math.sin(dlat / 2.0)
            + Math.cos(lat1Rad) * Math.cos(lat2Rad) * Math.sin(dlon / 2.0) * Math.sin(dlon / 2.0);
    double c = 2.0 * Math.asin(Math.sqrt(a));
    return RADIO_TIERRA_M * c;
  }

  /**
   * Heurística admisible h(n) para el A*: tiempo mínimo posible en segundos.
   *
   * <p>Espejo de {@code haversine_segundos(lat1, lon1, lat2, lon2)} Python. Calcula {@code
   * haversineM / V_MAX_MS}.
   *
   * @param lat1 latitud del punto 1 en grados decimales
   * @param lon1 longitud del punto 1 en grados decimales
   * @param lat2 latitud del punto 2 en grados decimales
   * @param lon2 longitud del punto 2 en grados decimales
   * @return tiempo mínimo posible en segundos
   */
  public static double haversineSegundos(double lat1, double lon1, double lat2, double lon2) {
    return haversineM(lat1, lon1, lat2, lon2) / V_MAX_MS;
  }
}
